assets/app/scripts/controllers/create.js

```diff
@@ -47,33 +119,198 @@ angular.module('openshiftConsole')
+      }
+    }
+
+    function filterCatalog() {
+      $scope.filteredCategoryTags = [];
+      $scope.filteredBuildersByCategory = {};
+      $scope.filteredTemplatesByCategory = {};
+      $scope.filteredNonBuilders = [];
+
+      angular.forEach($scope.categoryTags, function(tag) {
+        var builders = buildersByCategory[tag] || [],
+            templates = templatesByCategory[tag] || [],
+            filteredBuilders,
+            filteredTemplates;
+
+        filteredBuilders = builders.filter(function(builder) {
+          return matchesFilter(builder.name, builder.categoryTags);
+        });
+        $scope.filteredBuildersByCategory[tag] = filteredBuilders;
+
+        filteredTemplates = templates.filter(function(template) {
+          var tags = tagsFilter(template);
+          return matchesFilter(template.metadata.name, tags);
+        });
+        $scope.filteredTemplatesByCategory[tag] = filteredTemplates;
+
+        // Only add the category tag if there were any matches.
+        if (filteredBuilders.length || filteredTemplates.length) {
+          $scope.filteredCategoryTags.push(tag);
+        }
+      });
+
+      updateColumns();
+
+      $scope.filteredNonBuilders = nonBuilderImages.filter(function(image) {
+        return matchesFilter(image.name, image.categoryTags);
+      });
+    }
+
+    // Filter the catalog when the keyword or tag changes.
+    $scope.$watch('filter', filterCatalog, true);
+
+    function categorizeImages(imageStreams) {
+      angular.forEach(imageStreams, function(imageStream) {
+        if (!imageStream.status) {
+          return;
+        }
+
+        // Create a map of spec tags so we can find them efficiently later when
+        // looking at status tags.
+        var specTags = {};
+        if (imageStream.spec && imageStream.spec.tags) {
+          angular.forEach(imageStream.spec.tags, function(tag) {
+            if (tag.annotations && tag.annotations.tags) {
+              specTags[tag.name] = tag.annotations.tags.split(/\s*,\s*/);
+            }
+          });
         }
-      };
 
-      angular.forEach(projectTemplates, categorizeTemplates);
-      angular.forEach(openshiftTemplates, categorizeTemplates);
+        // Loop over status tags to categorize the images.
+        angular.forEach(imageStream.status.tags, function(tag) {
+          var imageStreamTag = tag.tag;
+          var category;
+          var categoryTags = specTags[imageStreamTag] || [];
+          var image = {
+            imageStream: imageStream,
+            imageStreamTag: imageStreamTag,
+            name: imageStream.metadata.name + ":" + imageStreamTag,
+            version: imageStreamTagAnnotationFilter(imageStream, 'version', imageStreamTag),
+            categoryTags: categoryTags
+          };
+          if (categoryTags.indexOf("builder") >= 0) {
+            // Add the builder image to its category.
+            category = getCategory(categoryTags);
+            if (!buildersByCategory[category]) {
```

Comment:
we usually make this a one-liner `buildersByCategory[category] = buildersByCategory[category] || [];`